### PR TITLE
When an error comes from loading the prelude, tell it happened at initialization time

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** enhancement

While working on the prelude, we may have errors generated at the time of (tacitly) requiring the `Init` library. In this case, the error has no location.

I wondered how to inform that the error comes from initialization and this PR is a rudimentary try at doing that. For instance, it might say:
```coq
While loading initial state:

Error:
Declaring a scope implicitly is deprecated; use in advance an explicit "Declare Scope list_scope.".
```

My naive approach has at least two problems:
- the rendering is not so nice; in particular, it adds an extra unwanted newline
- it does not work for warnings generated by loading the `Init` library

For the latter, we need a way to communicate to `cWarnings` that we are in the process of requiring `Init`. Would it make sense to see this information as a `loc`? Or shall it be an ad hoc information (e.g. a flag in `Flag` telling that we are in the initialization process?

### CODE REVIEW

File: stm/stm.ml [hunk 1]
```diff
@@ -2608,7 +2608,13 @@ let new_doc { doc_type ; iload_path; require_libs; stm_options } =
   end;
 
   (* Import initial libraries. *)
-  load_objs require_libs;
+  begin
+    try load_objs require_libs
+    with reraise when
```
Comment (gares) [inline stm/stm.ml]: I think you should `CErrors.push reraise` in order to keep the backtrace

**Review by gares — DISMISSED** [review]

**Comment by ejgallego** [comment]
IMHO the error should be catched by the caller of `new_doc`. In fact, there is code in `toplevel.ml` that supposedly should do that, however it doesn't get called for some reason [likely cruft].

I suggest to clean `toplevel.ml` up instead.

**Review by ejgallego — CHANGES_REQUESTED** [review]
Instead of modifying STM code to add ad-hoc error handling, refactor `toplevel.ml` so it has a better report and separation of concerns.

After the recent refactoring there, some responsibilities have been moved so indeed the precision on error reporting is worse now.

**Comment by herbelin** [comment]
> Instead of modifying STM code to add ad-hoc error handling, refactor toplevel.ml so it has a better report and separation of concerns.

Can you clarify please what you are expecting to be done in `toplevel.ml` (or `g_toplevel.ml4`?).

File: stm/stm.ml [hunk 2]
```diff
@@ -2608,7 +2608,13 @@ let new_doc { doc_type ; iload_path; require_libs; stm_options } =
...
+      (match reraise with Timeout -> true | e -> CErrors.noncritical e) ->
+         Feedback.msg_info (str "While loading initial state:");
+         raise reraise
+  end;
```
Comment (herbelin) [inline stm/stm.ml]: This is probably related to @ejgallego's comment, but I feel that it is wrong to print a `msg_info` here while this info message is supposed to be attached to the error message: no guarantee about what may happen inbetween (maybe some error handling components will decide not to display the error for instance, etc.). Both the extra newline and the question of how to make it available for warnings too suggest also that this is wrong.

**Comment by ejgallego** [comment]
> Can you clarify please what you are expecting to be done in toplevel.ml (or g_toplevel.ml4?).

so note that in:
```
      then begin
        let iload_path = build_load_path opts in
        let require_libs = require_libs opts in
        let stm_options = opts.stm_flags in
        try
          let open Vernac.State in
          let doc, sid =
            Stm.(new_doc
                   { doc_type = Interactive opts.toplevel_name;
                     iload_path; require_libs; stm_options;
                   }) in
          let state = { doc; sid; proof = None; time = opts.time } in
          Some (load_init_vernaculars opts ~state), opts
        with any -> flush_all(); fatal_error_exn any
      (* Non interactive: we perform a sequence of compilation steps *)
      end else begin
        try
          compile_files opts;
          (* Vio compile pass *)
          if opts.vio_files <> [] then schedule_vio opts;
          (* Vio task pass *)
          check_vio_tasks opts;
          (* Allow the user to output an arbitrary state *)
          outputstate opts;
          None, opts
        with any -> flush_all(); fatal_error_exn any
      end;
    with any ->
      flush_all();
      let extra = Some (str "Error during initialization: ") in
      fatal_error_exn ?extra any
  end in
  Feedback.del_feeder init_feeder;
  res
```
we have an inner `try ... catch ...` that exists. This is likely mimicking old code, however it prevents the outer catch to be reached and thus the right error message to be raised. I think indeed removing the first`try ... with ...` is the correct thing to do here, but we need to be careful, you know, this thing is tricky. In fact, I think the code should be refactored a bit deeper as to properly separate concerns.

Also, in the compilation phase, we create a document per file:
```
  match opts.compilation_mode with
  | BuildVo ->
      Flags.record_aux_file := true;
      let long_f_dot_v = ensure_v f_in in
      ensure_exists long_f_dot_v;
      let long_f_dot_vo =
        match f_out with
        | None -> long_f_dot_v ^ "o"
        | Some f -> ensure_vo long_f_dot_v f in

      let doc, sid = Stm.(new_doc
          { doc_type = VoDoc long_f_dot_vo;
            iload_path; require_libs; stm_options;
          }) in
```
that part indeed is badly wanting some refactoring too.

File: stm/stm.ml (hunk 2 above)
Comment (ejgallego) [reply stm/stm.ml]: Yup, I do agree.

**Comment by herbelin** [comment]
@ejgallego: thanks for the code in `coqtop.ml`. I would indeed remove the two inner `try` so that the proper message is printed. At least, that would indicate an error at initialization in interactive mode. (I pushed a commit removing them, thus forcing the printing of a preamble `"Error during initialization"` in this case.

About the original problem (in `coqc` mode), I don't see a trivial way to have both errors and warnings indicating that they come from initialization. As you say, there are several paths to locate an error or a warning:
- `Vernacentries` uses a `locate_if_not_already` on the `loc` of the vernac expr at the time it executes a command
- `cWarnings` uses a `current_loc` which is set in `Stm.add` and `Stm.query` before executing a command

Couldn't we already unite them? Does every command go through `Stm.add` or `Stm.query`? If yes, it could be located exclusively there?

Somehow, to implement the initial objective, the easiest way would be to put the information in a `Loc.t option` so that it can be extracted from the `current_loc` in `cWarnings.ml`, or added in `new_doc` whenever an error occurs.

But if we add this information to a `Loc.t option` it should be a third case on top of `None` and `Some` and we would lose the convenience of `?loc`. So, it would have to be in `Loc.t`, e.g. in the `fname`, but telling that the other fields are then irrelevant.

Hum...

**Comment by herbelin** [comment]
Well, I adopted a different approach then, but I'm walking on eggs. So, @gares, @ejgallego, I hope I'm not forgetting anything important.

- So that reporting errors at initialization time works for warnings, I added a new location source `Initialization`.
- So that the addition of a "toplevel" location works uniformly whether it is a warning or an error, and whether it is at initialization time or in a file or a toplevel input, I centralized the use of `CWarnings.set_current_loc` and `locate_if_not_located` in `process_transaction` (for interpretation) and `init_toplevel` (for initialization). Does it make sense?
- @ejgallego: I removed the nested `try` you were talking about in `coqtop.ml`.

(There are still a couple of unelegant things, but maybe not so important: I did not know how to add an info to an exception without raising it, so I modified instead `print_err_exn` so that it takes an `iexn`. Also, I did not know whether `locate_if_not_already` was better in `Loc` or in `CErrors`.)

PS: Apparently, amending a PR does not update the "code owners", so the primary maintainer for `lib`, for instance, has not been sollicitated even though `lib` is now changed.

**Comment by SkySkimmer** [comment]
>PS: Apparently, amending a PR does not update the "code owners", so the primary maintainer for lib, for instance, has not been sollicitated even though lib is now changed.

lib is @ejgallego who appears in the reviewer list to me.

**Comment by herbelin** [comment]
> lib is @ejgallego who appears in the reviewer list to me.

Sorry, I made a confusion with `clib`. 

I may have been also influenced in the background by the fact that this PR includes playing with adding info to exceptions, as well as playing with warnings, thus expecting at least @ppedrot to have been involved. On the other side, considering how reviewing is demanding, maybe better not to mandatorily add too many reviewers per PR.

**Comment by maximedenes** [comment]
> thus expecting at least @ppedrot to have been involved

What I usually do in these cases is to ping the corresponding experts with some questions on the code, but without assigning them formally as reviewers, if there are enough code owners looking at it.

**Comment by ejgallego** [comment]
I am not sure the change in location does make sense to me, but I need a few free minutes to review this PR, please bear with me @herbelin

File: lib/loc.ml [hunk 3]
```diff
@@ -13,6 +13,7 @@
 type source =
   | InFile of string
   | ToplevelInput
+  | Initialization
```
Comment (ejgallego) [inline lib/loc.ml]: I am positive this is not the way to go, IMHO locations should refer to file locations, what does the fact that system initialization has to do with where an object is located?

File: stm/stm.ml [hunk 4]
```diff
@@ -2924,9 +2925,12 @@ let process_transaction ?(newtip=Stateid.fresh ())
     in
     stm_pperr_endline (fun () -> str "processed with " ++ pr_rc rc ++ str " }}}");
     VCS.print ();
+    CWarnings.set_current_loc None;
     rc
   with e ->
     let e = CErrors.push e in
+    let e = Loc.locate_if_not_already ?loc e in
+    CWarnings.set_current_loc None;
```
Comment (ejgallego) [inline stm/stm.ml]: IMHO this `set_current_loc` is already a very ugly hack, let's not mess with is in this patch as it is orthogonal to the original purpose.

File: toplevel/coqargs.ml [hunk 5]
```diff
@@ -10,8 +10,9 @@
 
 let warning s = Flags.(with_option warn Feedback.msg_warning (Pp.strbrk s))
 
-let fatal_error ?extra exn =
-  Topfmt.print_err_exn ?extra exn;
+let fatal_error ?extra e =
+  Topfmt.print_err_exn ?extra e;
+  let exn, info = e in
```
Comment (ejgallego) [inline toplevel/coqargs.ml]: Why this change?

File: toplevel/coqargs.ml [hunk 6]
```diff
@@ -581,4 +582,6 @@ let parse_args arglist : coq_cmdopts * string list =
   in
   try
     parse init_args
-  with any -> fatal_error any
+  with any ->
+    let e = CErrors.push any in
+    fatal_error e
```
Comment (ejgallego) [inline toplevel/coqargs.ml]: Why this change?

File: toplevel/coqtop.ml [hunk 7]
```diff
@@ -469,34 +472,33 @@ let init_toplevel arglist =
...
+        let open Vernac.State in
+        let doc, sid =
+          Stm.(new_doc
+                 { doc_type = Interactive opts.toplevel_name;
+                   iload_path; require_libs; stm_options;
+                 }) in
+        let state = { doc; sid; proof = None; time = opts.time } in
+        Some (load_init_vernaculars opts ~state), opts
       (* Non interactive: we perform a sequence of compilation steps *)
       end else begin
```
Comment (ejgallego) [inline toplevel/coqtop.ml]: I think this is the part that has to be stripped out and that by the way has been causing quite a few other bugs; initialization is merged with a few different modes including vio and compilation, so no wonder the control flow has a lot to be desired; modes in the toplevel should be made clearer.

**Review by ejgallego — CHANGES_REQUESTED** [review]
I request 3 changes:
- don't add `Initialization` to `Loc` that simply is not right IMO.
- Don't mess with other parts of the system for what it should be a patch concentrated only in `coqtop`. I'd be happy to address other issues but in separate PRs.
- The real work to do here is to isolate the initialization routine out of the current mess of serial calls to effectful functions.

File: stm/stm.ml (hunk 4 above)
Comment (herbelin) [reply stm/stm.ml]: Warnings are printed in an asynchronous way. How to inform the warning printer of the current loc or of the current phase being interpreted.

One solution would be to have fully functional code and the current phase would be a field of the global state.

In the absence of a global state, how to avoid using a reference. Somehow, a bit better would be a `with_current_loc loc f` function but I fail to see how to do better. Do you have a suggestion?

File: lib/loc.ml (hunk 3 above)
Comment (herbelin) [reply lib/loc.ml]: I agree that a phase of an execution is not of the same nature as a location in a text. Locations are used in the step interpretation a sentence and are not used in the initialization step. Would a type of the form:
```ocaml
type initialization_step = Configuring | RequiringFile of dirpath | ...
type phase = Initialization of initialization_step | SentenceInterpretation of Location.t
```
be better for you?

Then, the warning printer and the main error printer would use this information to `locate_if_not_already` and/or add an appropriate phase message.

Or do you have something else in mind?

File: toplevel/coqargs.ml (hunk 5 above)
Comment (herbelin) [reply toplevel/coqargs.ml]: For bad reasons: I did not find how to add an info to an exception without reraising this exception. Otherwise said, how to build an `exn` from an `iexn`?

File: toplevel/coqargs.ml (hunk 6 above)
Comment (herbelin) [reply toplevel/coqargs.ml]: Same reason. I moved `fatal_error` to take an `iexn`, so I needed to turn an `exn` into a `iexn` (this step was previously done in `Topfmt.print_err_exn`.

PS: I'm learning and I'd be happy to know better how to deal with the `iexn` API.

File: stm/stm.ml (hunk 4 above)
Comment (ejgallego) [reply stm/stm.ml]: I agree, but this issue is orthogonal to the goal of this particular PR, so better to discuss changes to the warnings subsystem in its own issue/PR.

File: lib/loc.ml (hunk 3 above)
Comment (ejgallego) [reply lib/loc.ml]: IMHO the lower layers should know nothing about the phase, letting them know is just leakage, It is the top-level handler that has to be aware, so I don't see the point of introducing all such types. IMHO this PR is getting a bit complicated for what it should a 20 lines patch at most.

Comment (herbelin) [reply lib/loc.ml]: My intent with this PR was also to address the question of warnings.

For errors, I could move the information of the phase in `stm.ml` (would it be ok at the level of `process_transaction`?).

For warnings, I have to propagate the information one way or another (even if in another PR), and I don't see how to do it differently than by using a type such as `phase`...

Comment (ejgallego) [reply lib/loc.ml]: I see, I had looked at the original PR description which talks only about errors. So for the warnings, a few options are possible, but indeed this could be a good opportunity to get rid of the hideous set location hack and instead let `coqloop_feed` handle it.

I think the warnings change is fully orthogonal and other developers may want to get involved so I suggest we do it in a different PR and we get the fix for errors merged here.

**Comment by ejgallego** [comment]
> @ejgallego: Maybe it is already in your plans but it seems to me that in the topmost-level error printers (Stm.execution_error, Topfmt.print_err_exn, TopErr.print_error_for_buffer, ...?), we could enforce by type-checking that whenever we are interpreting a sentence, the loc is defined (and conversely, that when the loc is not defined, we are in an phase which is not evaluating a sentence, whether it is initialization or an internal phase of the stm if relevant).

IMHO these components should not care at all of what phase we are. For errors, the init error printer should know; for warnings, the feedback handler in `coqloop`.

File: stm/stm.ml (hunk 4 above)
Comment (herbelin) [reply stm/stm.ml]: OK, I think see: the `set_current_loc` could be removed once `coqloop_feed` is made aware of the phase.

For CoqIDE, is it relevant to transmit the "phase" or can we assume that there is always a loc? In particular, should we make the `slave_feeder` aware of the "phase"?

Comment (ejgallego) [reply stm/stm.ml]: The thing is that `CWarnings.set_current_loc` is totally broken, as soon as a command is parsed without the previous one being executed the info would be stale. IMHO, we cannot assume that there is a `loc`, it is the UI who has to handle that due to the possible asynchronous arrival of warnings. Moveover, the location in CoqIDE means little as it is relative to the current span. So IMHO, what the UI should do in the feeback handler if a message has no location is to use the `span_id` to locate the sentence and use that. This is in fact what's done in CoqIDE due to the relative nature of the locations there.

As the code stands today, initialization for CoqIDE will happen even before the CoqIDE-specific code is loaded, so I think no change should be necessary there for now.

**Comment by ejgallego** [comment]
@herbelin see #7194

**Comment by ejgallego** [comment]
Hugo (@herbelin ) please signal me when you consider this ready again by removing the tag, merci !

**Comment by herbelin** [comment]
@ejgallego: I'm failing to do better than (basically) the first version I proposed (this version was displaying an `Initialization` header for errors by acting at the level of `load_objs` of `new_doc`).

If I want that warnings raised while requiring the prelude are displayed also with a `During initialization` header, how I can do it? I need to communicate it to `TopErr.print_error_for_buffer`, but how?

So, unless you have an idea, I'm going to renounce to do better than the initial version. Sorry.

**Comment by ejgallego** [comment]
Sorry @herbelin I wasn't able to properly explain what I meant, I guess I am too much in a rush as of late.

I'll try to help with a more concrete explanation ASAP. A way to control the display better during the initialization phase is indeed to install a different feedback handler that takes a header.

I think that indeed the printers can already take some extra header.

I am positive that change to locations is IMHO not the proper one; thus keeping that request for changes.

**Comment by herbelin** [comment]
> install a different feedback handler that takes a header.

A handler as in `Feedback.add_feeder`? If yes, where do you store the header, we only have a `Feedback.feedback` as parameter? Do you store it in a global reference?

**Comment by herbelin** [comment]
@ejgallego: I think I guess the idea, but how to pass the informations? Typically, the change of feeder should surround the calls to `new_doc`. To do that, I need to know the id of the current feeder so as to be able to remove it and I need to know the mode of the current feeder so as to be able to reinstall it afterwards. So, that's new parameters to pass to the functions calling `new_doc`. Are you ready to add this?

On the other side, if we would pass this information, this would be easier to ensure that we have only one feeder registered at a time in the toplevel, so maybe should it be done?

**Comment by ejgallego** [comment]
Umm, I'm not sure I see what you mean, the initialization routine already uses its own feeder.

**Comment by herbelin** [comment]
@ejgallego: I pushed an attempt to set a specific feeder for each phase. Tell me if this is what you have in mind. (For the record, a simple version was at 08266d6945.)

**Comment by ejgallego** [comment]
This looks great thanks @herbelin , a quick look seems to indicate that this should be very close to something we can merge. Will review in detail in a couple of day, letting the CI run meanwhile.

**Comment by ejgallego** [comment]
Test suite fails.

**Comment by herbelin** [comment]
Test suite fixed.

**Comment by ejgallego** [comment]
@herbelin , c473c2d9c78d1245633b08433d707a8b1a5aa97c seems incorrect, doesn't it? In particular, you only enrich errors at document building time, but any error occurring in execution will now not be enriched, right?

**Comment by herbelin** [comment]
c473c2d  is actually a question. I don't know if the move makes more sense than before. As I said in the commit, the move excludes query and perform but includes more from `process_transaction`. I don't understand well enough in which context is each of these methods called to evaluate if the move brings something, if it loses something, or if it is just changes nothing.

In any case, `process_transaction` is called at each sentence, right?

**Comment by ejgallego** [comment]
> I propose to open a separate issue to discuss/fix this and remove that orthogonal change from this PR, as in particular I think it may make things worse indeed.

Pending on this.

**Comment by herbelin** [comment]
@ejgallego: I opened a separate issue (#7409) and remove the move of `locate_if_not_already` from this PR.

**Review by ejgallego — APPROVED** [review]
Looks pretty good, thanks @herbelin ; there is still some cleanup to do in the initalization procedure but this is a very good step IMO.
